infobind/cache.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infobind
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/sqlexec"
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultBindCacheSize = 5
+
+// BindData store the basic bind info and bindSql astNode.
+type BindData struct {
+	bindRecord
+	ast ast.StmtNode
+}
+
+// BindCache hold a bindDataMap, key:origin sql hash value: bindData slice.
+type BindCache struct {
+	Cache map[string][]*BindData
+}
+
+// Handle hold a atomic bindCache.
+type Handle struct {
+	bind atomic.Value
+}
+
+// HandleUpdater use to update the BindCache.
+type HandleUpdater struct {
+	Parser         *parser.Parser
+	LastUpdateTime types.Time
+	Ctx            sessionctx.Context
+	*Handle
+}
+
+type bindRecord struct {
+	OriginalSQL string
+	BindSQL     string
+	Db          string
+	Status      int64
+	CreateTime  types.Time
+	UpdateTime  types.Time
+	Charset     string
+	Collation   string
+}
+
+// NewHandle create a Handle with a BindCache.
+func NewHandle() *Handle {
+	handle := &Handle{}
+	return handle
+}
+
+// Get get bindCache from a Handle.
+func (h *Handle) Get() *BindCache {
+	bc := h.bind.Load()
+
+	if bc != nil {
+		return bc.(*BindCache)
+	}
+
```

Comment:
I think we can put the create new cache logic here.

Reply:
This function will be called by many thread, so we can't create cache in here. I will create new cache when first load all bind info in tidb

Reply:
Why we cannot create cache here in multi-thread case?

Reply:
If we create cache here in multi-thread case, this code will occur lazy initialize's race condition. So we need add a lock in Handle to resolve this problem.  I don't want to add lock in this code,

Reply:
You don't need to store the newly created cache here, just return it to the caller, and let the caller to store them.

Reply:
ok